lib/stellar/sep10.rb

```diff
@@ -0,0 +1,353 @@
+        value: value,
+        source_account: client
+      })
+
+      now = Time.now.to_i
+      tx.time_bounds = Stellar::TimeBounds.new(
+        min_time: now, 
+        max_time: now + timeout
+      )
+
+      tx.to_envelope(server).to_xdr(:base64)
+    end
+
+
+    Contract(C::KeywordArgs[
+      challenge: String,
+      server: Stellar::KeyPair
+    ] => [Stellar::TransactionEnvelope, String])    
+    # Reads a SEP 10 challenge transaction and returns the decoded transaction envelope and client account ID contained within.
+    #
+    # It also verifies that transaction is signed by the server.
+    #
+    # It does not verify that the transaction has been signed by the client or
+    # that any signatures other than the servers on the transaction are valid. Use
+    # one of the following functions to completely verify the transaction:
+    #    - Stellar::SEP10.verify_challenge_transaction_threshold
+    #    - Stellar::SEP10.verify_challenge_transaction_signers
+    #
+    # @param challenge [String] SEP0010 transaction challenge in base64.
+    # @param server [Stellar::KeyPair] keypair for server where the challenge was generated.
+    #
+    # @return [Stellar::TransactionEnvelope, String]
+    #
+    # = Example
+    # 
+    #   sep10 = Stellar::SEP10
+    #   challenge = sep10.build_challenge_tx(server: server, client: user, anchor_name: anchor, timeout: timeout) 
+    #   envelope, client_address = sep10.read_challenge_tx(challenge: challenge, server: server)
+    #
+    def self.read_challenge_tx(challenge:, server:)
+      envelope = Stellar::TransactionEnvelope.from_xdr(challenge, "base64") 
+      transaction = envelope.tx
+
+      if transaction.seq_num != 0
+        raise InvalidSep10ChallengeError.new(
+          "The transaction sequence number should be zero"
+        )
+      end
+
+      if transaction.source_account != server.public_key
+        raise InvalidSep10ChallengeError.new(
+          "The transaction source account is not equal to the server's account"
+        )
+      end
+
+      if transaction.operations.size != 1
+        raise InvalidSep10ChallengeError.new(
+          "The transaction should contain only one operation"
+        )
+      end
+
+      operation = transaction.operations.first
+      client_account_id = operation.source_account
+
+      if client_account_id.nil?
+        raise InvalidSep10ChallengeError.new(
+          "The transaction's operation should contain a source account"
+        )
+      end
+
+      if operation.body.arm != :manage_data_op
+        raise InvalidSep10ChallengeError.new(
+          "The transaction's operation should be manageData"
+        )
+      end
+
+      if operation.body.value.data_value.unpack("m")[0].size !=  48
+        raise InvalidSep10ChallengeError.new(
+          "The transaction's operation value should be a 64 bytes base64 random string"
+        )
+      end
+
+      if !verify_tx_signed_by(transaction_envelope: envelope, keypair: server)
```

Comment:
How does this know which network passphrase to verify with? The other SDKs accept the network passphrase as a parameter.

Reply:
So this relates to another issue with this SDK and the base repo. All methods that normally require a network passphrase but don't in the Ruby SDK use `Stellar.default_network`, which has to be assigned by the user. See [this issue](https://github.com/stellar/ruby-stellar-sdk/issues/22).

Reply:
My plan is to remove `Stellar.default_network` and add `network_passphrase` parameters to all the necessary functions. This is obviously a breaking change but there's a lot of breaking changes that will be unavoidable in order to get this SDK up to Horizon 1.0.